### ISSUE TITLE
please provide an API to get the entire expected and actual results

### ISSUE THREAD

**Issue by andyczerwonka** [xunit/xunit#1713]
In order for tool providers ([JetBrains Rider in this case](https://youtrack.jetbrains.com/issue/RIDER-4454#comment=27-2881108)) to provide diff tools, could you please provide an API to get the entire expected and actual results of a test?

**Comment by bradwilson** [comment]
Where and when is your code running that you're looking for this API?

**Comment by andyczerwonka** [comment]
As per the link in the request, the JetBrains Rider team would like to add a unit test diff tool similar to what they have in IntelliJ, but can't provide it without the API. See https://www.youtube.com/watch?v=R8l6D3qGSdg for an example of the tool working using JUnit.

**Comment by bradwilson** [comment]
That doesn't really answer my questions as to what kind of an API you'd expect, and to be called at what stage during the process. Are you post-processing our XML file? Are you writing a test runner? etc.

**Comment by andyczerwonka** [comment]
@gorohoroh, could you provide @bradwilson with some details on this please?

**Comment by gorohoroh** [comment]
I'd rather ask @AndreyAkinshin to provide any details and API expectations beyond what's covered in the [comment](https://youtrack.jetbrains.com/issue/RIDER-4454#comment=27-2881108).

**Comment by AndreyAkinshin** [comment]
@bradwilson, @andyczerwonka, @gorohoroh 
Sorry for late reply. We have own runner which directly use managed API. I guess, the best option will be extending of the `IFailureInformation` interface, it should provide additional metadata which can be used in Rider. Specifically, we need 3 additional fields: the full actual value, the full expected value, and the "assert kind" (e.g., `Equals`, `Contains`, and so on).

**Comment by vldF** [comment]
Any updates? 

I think this feature is really needed by the compiler and tooling community (I'm in the community btw). There is really common pattern when some input data like a code is being processed and the result artefact (like code in the target language, intermediate representation, etc) is being compared with the expected one. In this case you need to understand where is that hucking extra space or comma

**Comment by bradwilson** [comment]
@vldF This is not currently on our backlog.

**Comment by bradwilson** [comment]
I am prototyping the changes for this today.

I assume this would be a sufficient addition to the base `XunitException`:

```diff
+        /// <summary>
+        /// Gets the raw actual value.
+        /// </summary>
+        /// <remarks>
+        /// Not all assertions have an "actual" value, so a <c>null</c> result may mean either a <c>null</c>
+        /// value was passed for actual, or it may mean there is no actual value. Using the context of
+        /// <see cref="AssertionName"/> is required to know when there may or may not be an actual value.
+        /// </remarks>
+        public object? Actual { get; }
+
+        /// <summary>
+        /// Gets the name of the assertion that caused this exception.
+        /// </summary>
+        /// <remarks>
+        /// The name of the assertion may not be known, or the exception may not have resulted from an
+        /// assertion failure. In that case, the special value <c>Unknown</c> may be returned. In addition,
+        /// when there are non-async and async overloads (for example, <c>All</c> vs. <c>AllAsync</c>),
+        /// you should standardize to send both using the simpler name (in this case, <c>All</c>).
+        /// </remarks>
+        public string AssertionName { get; }
+
+        /// <summary>
+        /// Gets the raw expected value.
+        /// </summary>
+        /// <remarks>
+        /// Not all assertions have an "expected" value, so a <c>null</c> result may mean either a <c>null</c>
+        /// value was passed for expected, or it may mean there is no expected value. Using the context of
+        /// <see cref="AssertionName"/> is required to know when there may or may not be an expected value.
+        /// </remarks>
+        public object? Expected { get; }
```

Correct?

**Comment by bradwilson** [comment]
Going through the motions for this, there is a blocking issue.

xUnit.net v3 projects live in their own executable space. That means any exception reported back has to be serialized and deserialized on the other side. There is no generalized serialization system for arbitrary values that crosses process boundaries.

I don't think this is possible in v3.

**Comment by bradwilson** [comment]
For a more complete answer:

The cross-process requirement arises from the VSTest APIs. The VSTest adapter (`xunit.runner.visualstudio`) lives in the process space of the VSTest host process. When running a v3 test project, it is launched into its own process space, and the communication mechanism back to the host process is via serialized xUnit.net status messages. Today, the form of those is designed entirely so that their contents can be serialized back and forth via JSON.

Every error-containing message (including `ITestFailed`) implements [`IErrorMetadata`](https://github.com/xunit/xunit/blob/72878584ab032c9d7cd02a3387de36e1b3233938/src/xunit.v3.common/Abstractions/Messages/Metadata/IErrorMetadata.cs). Implementing these three new properties would require use to have JSON-compatible values for `Expected` and `Actual` (`AssertionName` is already JSON compatible). That means the values need to be serialized.

xUnit.net has its own built in serialization system (via [`IXunitSerializable`](https://github.com/xunit/xunit/blob/72878584ab032c9d7cd02a3387de36e1b3233938/src/xunit.v3.common/Abstractions/Serialization/IXunitSerializable.cs) and [`IXunitSerializer`](https://github.com/xunit/xunit/blob/72878584ab032c9d7cd02a3387de36e1b3233938/src/xunit.v3.common/Abstractions/Serialization/IXunitSerializer.cs)). These types exist today because there is no generalized arbitrary object serialization mechanism that is suitable. The closest one would be `DataContractSerializer` by way of `ISerializable`, which is not well supported by types in .NET.

In short, to implement this feature it would require someone to write custom serializers for everything they would ever have placed into an assertion, which is an incredibly high bar that I think nobody would be interested in trying to climb.

**Comment by vldF** [comment]
Hmm, it seems I fixed the issue this way:

```
import protokt.v1.gradle.ProtoktPlugin
import protokt.v1.gradle.protokt

plugins {
    kotlin("jvm") version "2.0.0"
    id("com.toasttab.protokt") version "1.0.0-beta.3" apply false
}

group = "com.example"
version = "1.0-SNAPSHOT"

kotlin {
    jvmToolchain(17)
}

tasks.withType<JavaCompile> {
    enabled = false
}

dependencies {
    implementation("com.google.protobuf:protobuf-java:4.29.1")
//    implementation("com.toasttab.protokt:protokt-runtime-grpc-jvm:1.0.0-beta.3")
    implementation("io.grpc:grpc-protobuf:1.64.0")
    implementation("io.grpc:grpc-kotlin-stub:1.4.1")
}

allprojects {
    repositories {
        mavenCentral()
        gradlePluginPortal()
        mavenLocal()
    }
}

apply<ProtoktPlugin>()

protokt {
    generate {
        all()
    }
}
```

I just apply the protokt plugin at the build.gradle.kts end. Is it ok?
